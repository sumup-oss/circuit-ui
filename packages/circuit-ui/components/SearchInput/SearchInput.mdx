import { Meta, Status, Props, Story } from '../../../../.storybook/components';
import * as Stories from './SearchInput.stories';

<Meta of={Stories} />

# SearchInput

<Status.Stable />

Search Input is an input that indicates to the user they can use the field
to search information on the screen.

<Story of={Stories.Base} />
<Props />

## Component variations

## Disabled

The search input can be disabled.

<Story of={Stories.Disabled} />

## Clearable

A small 'x' appears once a users starts typing in the field. When clicked, the button clears the field.

<Story of={Stories.Clearable} />
