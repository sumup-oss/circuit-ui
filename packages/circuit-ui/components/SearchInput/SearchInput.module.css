.base::-webkit-search-decoration,
.base::-webkit-search-cancel-button,
.base::-webkit-search-results-button,
.base::-webkit-search-results-decoration {
  appearance: none;
}
