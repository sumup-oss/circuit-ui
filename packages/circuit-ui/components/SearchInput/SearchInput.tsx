/**
 * Copyright 2019, SumUp Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { forwardRef } from 'react';
import { css } from '@emotion/core';
import { Search, Close } from '@sumup/icons';

import styled, { StyleProps } from '../../styles/styled';
import Input from '../Input';
import { InputProps } from '../Input/Input';
import IconButton from '../IconButton';

type ClearProps =
  | { onClear?: never; clearLabel?: never }
  | {
      /**
       * Callback function when the user clears the field.
       */
      onClear: () => void;
      /**
       * Visually hidden text label on the clear button for screen readers.
       * Crucial for accessibility.
       */
      clearLabel: string;
    };

export type SearchInputProps = InputProps & ClearProps;

const clearButtonStyles = ({ theme }: StyleProps) => css`
  border: none;
  margin: 0 !important;
  width: auto !important;
  pointer-events: all !important;
  cursor: pointer !important;
  border-radius: ${theme.borderRadius.byte};
`;

const ClearButton = styled(IconButton)(clearButtonStyles);

/**
 * SearchInput component for forms.
 */
export const SearchInput = forwardRef(
  (
    { value, onClear, clearLabel, ...props }: SearchInputProps,
    ref: SearchInputProps['ref'],
  ) => {
    if (
      process.env.UNSAFE_DISABLE_ACCESSIBILITY_ERRORS !== 'true' &&
      process.env.NODE_ENV !== 'production' &&
      process.env.NODE_ENV !== 'test' &&
      onClear &&
      !clearLabel
    ) {
      throw new Error(
        'The SearchInput component is missing a `clearLabel` prop. This is an accessibility requirement. Omit the `onClear` prop if you intend to disable the input clearing functionality.',
      );
    }
    return (
      <Input
        value={value}
        type="text"
        renderPrefix={(renderProps) => <Search size="16" {...renderProps} />}
        renderSuffix={(renderProps) =>
          value && onClear && clearLabel ? (
            <ClearButton onClick={onClear} label={clearLabel} {...renderProps}>
              <Close size="16" />
            </ClearButton>
          ) : null
        }
        {...props}
        ref={ref}
      />
    );
  },
);

SearchInput.displayName = 'SearchInput';
