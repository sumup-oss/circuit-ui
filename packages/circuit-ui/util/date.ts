/**
 * Copyright 2024, SumUp Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Temporal } from 'temporal-polyfill';
import { formatDateTime } from '@sumup-oss/intl';

import type { Locale } from './i18n.js';

export type FirstDayOfWeek = 1 | 7;
export type DaysInWeek = number;
export type PlainDateRange =
  | []
  | [Temporal.PlainDate]
  | [Temporal.PlainDate, Temporal.PlainDate];

// ISO 8601 timestamps only support positive 4-digit years
export const MIN_YEAR = 1;
export const MAX_YEAR = 9999;

export const MIN_MONTH = 1;
export const MAX_MONTH = 12;

export const MIN_DAY = 1;
// MAX_DAY is not constant as it depends on the year and month

export function getTodaysDate() {
  return Temporal.Now.plainDateISO();
}

export function isPlainDate(date: unknown): date is Temporal.PlainDate {
  return date instanceof Temporal.PlainDate;
}

export function toPlainDate(date?: string): Temporal.PlainDate | undefined {
  if (!date) {
    return undefined;
  }
  try {
    return Temporal.PlainDate.from(date);
  } catch {
    return undefined;
  }
}

export function sortDateRange<T extends PlainDateRange>(dateRange: T): T {
  return dateRange.sort((a, b) => Temporal.PlainDate.compare(a, b)) as T;
}

export function clampDate(
  date: Temporal.PlainDate,
  minDate?: Temporal.PlainDate | null,
  maxDate?: Temporal.PlainDate | null,
): Temporal.PlainDate {
  if (minDate && Temporal.PlainDate.compare(date, minDate) < 0) {
    return minDate;
  }
  if (maxDate && Temporal.PlainDate.compare(date, maxDate) > 0) {
    return maxDate;
  }
  return date;
}

export function getFirstDateOfWeek(
  date: Temporal.PlainDate,
  firstDayOfWeek: FirstDayOfWeek,
): Temporal.PlainDate {
  const diff =
    (date.dayOfWeek < firstDayOfWeek ? date.daysInWeek : 0) +
    date.dayOfWeek -
    firstDayOfWeek;
  return date.subtract({ days: diff });
}

export function getLastDateOfWeek(
  date: Temporal.PlainDate,
  firstDayOfWeek: FirstDayOfWeek,
): Temporal.PlainDate {
  return getFirstDateOfWeek(date, firstDayOfWeek).add({
    days: date.daysInWeek - 1,
  });
}

export function getMonthName(month: number, locale?: Locale) {
  // The year can be arbitrary since the month names are the same every year
  const yearMonth = new Temporal.PlainYearMonth(2000, month, 'gregory');
  return formatDateTime(yearMonth, locale, { month: 'long' });
}
