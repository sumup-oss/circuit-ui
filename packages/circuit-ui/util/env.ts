/**
 * Copyright 2024, SumUp Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

export function getEnvVariable(name: string) {
  if (name === 'NODE_ENV') {
    // Some bundlers have special logic for `process.env.NODE_ENV` which
    // relies on it being written as a continuous string. Destructuring or
    // dynamically accessing `NODE_ENV` can break this logic.
    throw new Error('Do not dynamically access NODE_ENV');
  }

  if (typeof process !== 'undefined') {
    return process.env?.[name];
  }

  return undefined;
}
