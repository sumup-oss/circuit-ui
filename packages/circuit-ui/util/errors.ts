/**
 * Copyright 2022, SumUp Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/* eslint-disable max-classes-per-file */
import React from 'react';

export class CircuitError extends Error {
  constructor(componentName: string, message: string) {
    super(`[${componentName}] ${message}`);
    this.name = 'CircuitError';
    // Adapted from https://stackoverflow.com/questions/33474179/react-access-parent-component-name

    /* eslint-disable max-len, no-underscore-dangle, @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-unsafe-call */
    if (
      // @ts-expect-error Since this code only runs in development, it's fine to use this internal React API.
      React.__SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED
        .ReactCurrentOwner &&
      // @ts-expect-error This is fine.
      React.__SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED.ReactCurrentOwner
        .return &&
      // @ts-expect-error This is fine.
      React.__SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED.ReactCurrentOwner
        .return.type
    ) {
      this.stack =
        // @ts-expect-error This is fine.
        React.__SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED
          .ReactCurrentOwner.return.type as string;
    }
    if (
      // @ts-expect-error Since this code only runs in development, it's fine to use this internal React API.
      React.__SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED
        .ReactDebugCurrentFrame &&
      // @ts-expect-error This is fine.
      React.__SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED
        .ReactDebugCurrentFrame.getCurrentStack
    ) {
      this.stack =
        // @ts-expect-error This is fine.
        React.__SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED.ReactDebugCurrentFrame.getCurrentStack() as string;
    }
    /* eslint-enable max-len, no-underscore-dangle, @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-unsafe-call */
  }
}

export class DeprecationError extends CircuitError {
  constructor(componentName: string, message: string) {
    super(componentName, message);
    this.name = 'DeprecationError';
  }
}

export class AccessibilityError extends CircuitError {
  constructor(componentName: string, message: string) {
    super(componentName, message);
    this.name = 'AccessibilityError';
  }
}
