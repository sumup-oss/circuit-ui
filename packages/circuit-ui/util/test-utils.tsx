/**
 * Copyright 2020, SumUp Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { FunctionComponent, ReactElement } from 'react';
import { renderToStaticMarkup } from 'react-dom/server';
import '@testing-library/jest-dom/extend-expect';
import { configureAxe } from 'jest-axe';
import { render as renderTest, RenderResult } from '@testing-library/react';
import { renderHook, act as actHook } from '@testing-library/react-hooks';
import userEvent from '@testing-library/user-event';
import { ThemeProvider } from 'emotion-theming';
import { light } from '@sumup/design-tokens';

import {
  ComponentsContext,
  defaultComponents,
} from '../components/ComponentsContext/ComponentsContext';

export * from '@testing-library/react';

export type RenderFn<T = any> = (component: ReactElement, ...rest: any) => T;

const WithProviders: FunctionComponent = ({ children }) => (
  <ComponentsContext.Provider value={defaultComponents}>
    <ThemeProvider theme={light}>{children}</ThemeProvider>
  </ComponentsContext.Provider>
);

const render: RenderFn<RenderResult> = (component, options) =>
  renderTest(component, { wrapper: WithProviders, ...options });
const renderToHtml: RenderFn<string> = (component) =>
  renderToStaticMarkup(<WithProviders>{component}</WithProviders>);
const create = (...args: Parameters<RenderFn<RenderResult>>) => {
  const { container } = render(...args);
  return container.children.length > 1
    ? container.children
    : container.firstChild;
};

const axe = configureAxe({
  rules: {
    // disabled landmark rules when testing isolated components.
    region: { enabled: false },
  },
});

export { create, render, renderToHtml, renderHook, actHook, userEvent, axe };
