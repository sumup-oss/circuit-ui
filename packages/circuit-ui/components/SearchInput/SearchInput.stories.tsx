/**
 * Copyright 2019, SumUp Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React, { useState, ChangeEvent } from 'react';

import { SearchInput, SearchInputProps } from './SearchInput';
import docs from './SearchInput.docs.mdx';

export default {
  title: 'Forms/Input/SearchInput',
  component: SearchInput,
  parameters: {
    docs: { page: docs },
  },
  argTypes: {
    placeholder: { control: 'text' },
  },
};

const baseArgs = {
  placeholder: 'Type a word...',
  label: 'Search',
};

export const Base = (args: SearchInputProps) => <SearchInput {...args} />;

Base.args = baseArgs;

export const Disabled = (args: SearchInputProps) => <SearchInput {...args} />;

Disabled.args = { ...baseArgs, disabled: true };

export const Clearable = (args: SearchInputProps) => {
  const [value, setValue] = useState('');

  const handleChange = ({
    target: { value: inputValue },
  }: ChangeEvent<HTMLInputElement>) => {
    setValue(inputValue);
  };

  const handleClear = () => {
    setValue('');
  };

  return (
    <SearchInput
      {...args}
      value={value}
      onClear={handleClear}
      onChange={handleChange}
    />
  );
};

Clearable.args = baseArgs;
