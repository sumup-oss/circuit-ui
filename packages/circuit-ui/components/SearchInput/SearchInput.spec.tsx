/**
 * Copyright 2019, SumUp Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { createRef } from 'react';

import { create, render, renderToHtml, axe } from '../../util/test-utils';

import SearchInput from '.';

describe('SearchInput', () => {
  const baseProps = { label: 'Search' };

  /**
   * Style tests.
   */
  it('should render with default styles', () => {
    const actual = create(<SearchInput {...baseProps} />);
    expect(actual).toMatchSnapshot();
  });

  it('should grey out icon when disabled', () => {
    const actual = create(<SearchInput {...baseProps} disabled />);
    expect(actual).toMatchSnapshot();
  });

  it('should display a clear icon when not empty and an onClear callback is provided', () => {
    const mockCallback = jest.fn();
    const clearLabel = 'Clear';

    const { getByRole } = render(
      <SearchInput
        {...baseProps}
        value="Search value"
        clearLabel={clearLabel}
        onClear={mockCallback}
        /**
         * We set onChange to silence a warning about adding a `value` without
         * `onChange` or `readOnly`.
         */
        onChange={mockCallback}
      />,
    );
    expect(getByRole('button')).toBeVisible();
    expect(getByRole('button')).toHaveTextContent(clearLabel);
  });

  describe('business logic', () => {
    /**
     * Should accept a working ref
     */
    it('should accept a working ref', () => {
      const tref = createRef<HTMLInputElement & HTMLTextAreaElement>();
      const { container } = render(<SearchInput {...baseProps} ref={tref} />);
      const input = container.querySelector('input');
      expect(tref.current).toBe(input);
    });
  });

  /**
   * Accessibility tests.
   */
  it('should meet accessibility guidelines', async () => {
    const wrapper = renderToHtml(<SearchInput {...baseProps} />);
    const actual = await axe(wrapper);
    expect(actual).toHaveNoViolations();
  });
});
