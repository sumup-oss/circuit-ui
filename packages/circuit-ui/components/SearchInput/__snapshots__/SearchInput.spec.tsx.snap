// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchInput should grey out icon when disabled 1`] = `
.circuit-0 {
  opacity: 0.5;
  pointer-events: none;
  box-shadow: none;
}

.circuit-1 {
  display: block;
  font-size: 14px;
  line-height: 20px;
}

.circuit-2 {
  display: inline-block;
  margin-bottom: 4px;
}

.circuit-4 {
  position: relative;
}

.circuit-5 {
  position: absolute;
  pointer-events: none;
  color: #666;
  padding: 12px 16px;
  height: 48px;
  width: 48px;
}

.circuit-6 {
  font-size: 16px;
  line-height: 24px;
  -webkit-appearance: none;
  background-color: #FFF;
  border: none;
  outline: 0;
  border-radius: 8px;
  padding: 12px 16px;
  -webkit-transition: box-shadow 120ms ease-in-out,padding 120ms ease-in-out;
  transition: box-shadow 120ms ease-in-out,padding 120ms ease-in-out;
  width: 100%;
  margin: 0;
  background-color: #E6E6E6;
  padding-left: 48px;
  padding-right: 48px;
  box-shadow: 0 0 0 1px #999;
}

.circuit-6::-webkit-input-placeholder {
  color: #999;
  -webkit-transition: color 120ms ease-in-out;
  transition: color 120ms ease-in-out;
}

.circuit-6::-moz-placeholder {
  color: #999;
  -webkit-transition: color 120ms ease-in-out;
  transition: color 120ms ease-in-out;
}

.circuit-6:-ms-input-placeholder {
  color: #999;
  -webkit-transition: color 120ms ease-in-out;
  transition: color 120ms ease-in-out;
}

.circuit-6::placeholder {
  color: #999;
  -webkit-transition: color 120ms ease-in-out;
  transition: color 120ms ease-in-out;
}

<div
  class="circuit-0"
>
  <label
    class="circuit-1"
    for="input_2"
  >
    <span
      class="circuit-2 circuit-3"
    >
      Search
    </span>
    <div
      class="circuit-4"
    >
      <svg
        class="circuit-5"
        fill="none"
        height="16"
        viewBox="0 0 16 16"
        width="16"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M12 7A5 5 0 1 1 2 7a5 5 0 0 1 10 0zm-.81 5.608a7 7 0 1 1 1.418-1.418l3.102 3.103c.186.189.29.443.29.709a.999.999 0 0 1-.998.998c-.266 0-.52-.104-.71-.29l-3.102-3.102zM14 5.238v-.472c.004.157.004.315 0 .472z"
          fill="currentColor"
          fill-rule="evenodd"
        />
      </svg>
      <input
        class="circuit-6"
        disabled=""
        id="input_2"
        type="text"
        value=""
      />
    </div>
  </label>
</div>
`;

exports[`SearchInput should render with default styles 1`] = `
.circuit-1 {
  display: block;
  font-size: 14px;
  line-height: 20px;
}

.circuit-2 {
  display: inline-block;
  margin-bottom: 4px;
}

.circuit-4 {
  position: relative;
}

.circuit-5 {
  position: absolute;
  pointer-events: none;
  color: #666;
  padding: 12px 16px;
  height: 48px;
  width: 48px;
}

.circuit-6 {
  font-size: 16px;
  line-height: 24px;
  -webkit-appearance: none;
  background-color: #FFF;
  border: none;
  outline: 0;
  border-radius: 8px;
  padding: 12px 16px;
  -webkit-transition: box-shadow 120ms ease-in-out,padding 120ms ease-in-out;
  transition: box-shadow 120ms ease-in-out,padding 120ms ease-in-out;
  width: 100%;
  margin: 0;
  padding-left: 48px;
  padding-right: 48px;
  box-shadow: 0 0 0 1px #999;
}

.circuit-6::-webkit-input-placeholder {
  color: #999;
  -webkit-transition: color 120ms ease-in-out;
  transition: color 120ms ease-in-out;
}

.circuit-6::-moz-placeholder {
  color: #999;
  -webkit-transition: color 120ms ease-in-out;
  transition: color 120ms ease-in-out;
}

.circuit-6:-ms-input-placeholder {
  color: #999;
  -webkit-transition: color 120ms ease-in-out;
  transition: color 120ms ease-in-out;
}

.circuit-6::placeholder {
  color: #999;
  -webkit-transition: color 120ms ease-in-out;
  transition: color 120ms ease-in-out;
}

.circuit-6:hover {
  box-shadow: 0 0 0 1px #666;
}

.circuit-6:focus {
  box-shadow: 0 0 0 2px #3063E9;
}

.circuit-6:active {
  box-shadow: 0 0 0 1px #3063E9;
}

<div
  class="circuit-0"
>
  <label
    class="circuit-1"
    for="input_1"
  >
    <span
      class="circuit-2 circuit-3"
    >
      Search
    </span>
    <div
      class="circuit-4"
    >
      <svg
        class="circuit-5"
        fill="none"
        height="16"
        viewBox="0 0 16 16"
        width="16"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M12 7A5 5 0 1 1 2 7a5 5 0 0 1 10 0zm-.81 5.608a7 7 0 1 1 1.418-1.418l3.102 3.103c.186.189.29.443.29.709a.999.999 0 0 1-.998.998c-.266 0-.52-.104-.71-.29l-3.102-3.102zM14 5.238v-.472c.004.157.004.315 0 .472z"
          fill="currentColor"
          fill-rule="evenodd"
        />
      </svg>
      <input
        class="circuit-6"
        id="input_1"
        type="text"
        value=""
      />
    </div>
  </label>
</div>
`;
